### PR TITLE
optimize FirstOutstanding in the sent packet history

### LINKED ISSUE #3462: Optimizing (*sentPacketHistory).FirstOutstanding
Hi @marten-seemann,

quic-go currently spends a lot of CPU time on `(*sentPacketHistory).FirstOutstanding` when transmitting at high speed over a high packet loss connection. (Of course, usually this can't happen at all, as the built-in CUBIC congestion control slows the connection down to a crawl when packet loss rate is high. I have replaced it with a custom constant-speed CC.)

The following call graph was generated by a transmission test between 2 servers ($5 AWS Lightsail) @ 200 Mbps, simulated 30ms, 4% packet loss link

profile001

image

```
Showing nodes accounting for 3080ms, 28.33% of 10870ms total
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context
----------------------------------------------------------+-------------
                                            1640ms 53.25% |   github.com/lucas-clemente/quic-go/internal/ackhandler.(*sentPacketHistory).HasOutstandingPackets (inline)
                                            1440ms 46.75% |   github.com/lucas-clemente/quic-go/internal/ackhandler.(*sentPacketHandler).PeekPacketNumber (inline)
    2490ms 22.91% 22.91%     3080ms 28.33%                | github.com/lucas-clemente/quic-go/internal/ackhandler.(*sentPacketHistory).FirstOutstanding
                                             590ms 19.16% |   github.com/lucas-clemente/quic-go/internal/ackhandler.(*PacketElement).Next (inline)
----------------------------------------------------------+-------------
                                             590ms   100% |   github.com/lucas-clemente/quic-go/internal/ackhandler.(*sentPacketHistory).FirstOutstanding (inline)
     590ms  5.43% 28.33%      590ms  5.43%                | github.com/lucas-clemente/quic-go/internal/ackhandler.(*PacketElement).Next
----------------------------------------------------------+-------------
```

`FirstOutstanding` was taking up nearly 30% of the runtime and was a major CPU bottleneck. The program occupied the entire CPU core, couldn't get faster than a merely 100 Mbps because of it. And it gets worse as I crank up the loss rate. Switching to servers with better CPU performance does improve the situation - I was able to hit ~700 Mbps between two Ryzen 7 machines over the same link.

Do you have any thoughts on what could be done to optimize this part of the code? Perhaps we can add some kind of cache, or pre-calculate the first outstanding packet every time the linked list is modified, instead of iterating it every call?

pprof sample attached:
[pprof.hy_linux_amd64.samples.cpu.002.pb.gz](https://github.com/lucas-clemente/quic-go/files/9006604/pprof.hy_linux_amd64.samples.cpu.002.pb.gz)

### PR DESCRIPTION
**Pull request by tobyxdd**
Closes #3462

### CODE REVIEW

**Comment by tobyxdd** [comment]
Working on the integration test failures

**Comment by tobyxdd** [comment]
@marten-seemann I need some help 🤨 still haven't figured out the cause yet

**Comment by tobyxdd** [comment]
nvm, found the problem.

File: internal/ackhandler/sent_packet_history.go [hunk 1]
```diff
@@ -81,28 +101,49 @@ func (h *sentPacketHistory) Remove(p protocol.PacketNumber) error {
...
 func (h *sentPacketHistory) HasOutstandingPackets() bool {
-	return h.FirstOutstanding() != nil
+	return h.outstandingPacketList.Len() > 0
 }
 
 func (h *sentPacketHistory) DeleteOldPackets(now time.Time) {
 	maxAge := 3 * h.rttStats.PTO(false)
 	var nextEl *PacketElement
-	for el := h.packetList.Front(); el != nil; el = nextEl {
+	for el := h.etcPacketList.Front(); el != nil; el = nextEl {
```
Comment (marten-seemann) [inline internal/ackhandler/sent_packet_history.go:117]: Can you add a comment here why we don't have to consider the `oustandingPacketList`?

File: internal/ackhandler/sent_packet_history.go [hunk 2]
```diff
@@ -81,28 +101,49 @@ func (h *sentPacketHistory) Remove(p protocol.PacketNumber) error {
...
+	}
+}
+
+func (h *sentPacketHistory) DeclareLost(p *Packet) *Packet {
+	el, ok := h.packetMap[p.PacketNumber]
+	if !ok {
+		return nil
+	}
+	h.outstandingPacketList.Remove(el)
+	h.etcPacketList.Remove(el)
```
Comment (marten-seemann) [inline internal/ackhandler/sent_packet_history.go:134]: Removing from both lists doesn't seem very clean. Should we save in the map which list we've stored it in?

File: internal/ackhandler/sent_packet_history.go [hunk 3]
```diff
@@ -61,15 +82,14 @@ func (h *sentPacketHistory) Iterate(cb func(*Packet) (cont bool, err error)) err
...
 func (h *sentPacketHistory) FirstOutstanding() *Packet {
-	for el := h.packetList.Front(); el != nil; el = el.Next() {
-		p := &el.Value
-		if !p.declaredLost && !p.skippedPacket && !p.IsPathMTUProbePacket {
-			return p
-		}
+	el := h.outstandingPacketList.Front()
+	if el == nil {
+		return nil
+	} else {
```
Comment (marten-seemann) [inline internal/ackhandler/sent_packet_history.go:90]: Same here.

**Review by marten-seemann — COMMENTED** [review]
Thank you! This looks pretty good.

It would probably be good to have a test that `Iterate` iterates in the right order, when skipping between the two lists.

File: internal/ackhandler/sent_packet_history.go (hunk 2 above)
Comment (tobyxdd) [reply internal/ackhandler/sent_packet_history.go:136]: `Remove` already checks whether `e.list == l`. Do we really need to add an extra step?

**Comment by tobyxdd** [comment]
> It would probably be good to have a test that `Iterate` iterates in the right order, when skipping between the two lists.

I think the tests involving `expectInPacketHistory` already cover it.

**Comment by tobyxdd** [comment]
> It would probably be good to have a test that `Iterate` iterates in the right order, when skipping between the two lists.

done

**Comment by tobyxdd** [comment]
@marten-seemann the latest commit https://github.com/lucas-clemente/quic-go/pull/3467/commits/7cedeed64a5bd52ac00e84249b33c8415b26a061 tries to solve https://github.com/lucas-clemente/quic-go/issues/300 using AVL tree for gaps.

I've already done most of the work but there's a nasty bug somewhere in handling the start/end gaps causing the tests to fail. I spent the entire night trying to debug without any luck... maybe you can take a look?

**Comment by tobyxdd** [comment]
I've found the problem and fixed it. Can confirm that it works perfectly in real-world applications too.

With these 2 changes I can get a stable 400-500 Mbps on $5 AWS Lightsail - more than quadrupling the performance.

> Can you please separate this into a separate branch / PR though?

Maybe we can just put them together as one optimization PR?

**Comment by marten-seemann** [comment]
Where’s that tree implementation coming from? I’m not convinced that quic-go should be in the business of implementing a tree data structure from scratch.

**Comment by marten-seemann** [comment]
> Or we can just put them together as one big optimization PR?

These are two completely different issues. They need to be resolved in separate PRs.

**Comment by tobyxdd** [comment]
> Where’s that tree implementation coming from? I’m not convinced that quic-go should be in the business of implementing a tree data structure from scratch.

It's from https://github.com/ross-oreto/go-tree/blob/master/btree.go but I've made various modifications to meet our use case

**Comment by tobyxdd** [comment]
frameSorter changes removed

File: internal/ackhandler/sent_packet_history.go (hunk 2 above)
Comment (marten-seemann) [reply internal/ackhandler/sent_packet_history.go:136]: Ok, given that this introduces a bunch of complexity, we probably don't need it.

Can you add a comment that `Remove` checks if the element is contained in the list, and is a no-op if it isn't.

**Review by marten-seemann — APPROVED** [review]
This looks good, thanks for your work @tobyxdd!

One tiny request for a comment, then we're ready to merge.

**Review by marten-seemann — APPROVED** [review]
Thank you!

**Pull request merged**